src/server/pkg/storage/fileset/merge.go

```diff
@@ -31,64 +31,80 @@ func newMergeReader(rs []*Reader) *MergeReader {
 	return &MergeReader{pq: newPriorityQueue(fileStreams)}
 }
 
-func (mr *MergeReader) iterate(f func(*FileMergeReader) error) error {
+func (mr *MergeReader) iterate(cb func(*FileMergeReader) error) error {
 	return mr.pq.iterate(func(ss []stream, _ ...string) error {
-		// Convert generic streams to file streams.
-		var fss []*fileStream
-		for _, s := range ss {
-			fss = append(fss, s.(*fileStream))
-		}
-		// Create file merge reader and execute callback.
-		var frs []*FileReader
-		for _, fs := range fss {
-			fr, err := fs.r.Next()
-			if err != nil {
-				return err
-			}
-			frs = append(frs, fr)
-		}
-		frs = applyFullDeletes(frs)
-		contentTags := computeContentTags(frs)
-		if len(contentTags) == 0 {
-			return nil
-		}
-		return f(newFileMergeReader(frs, sizeOfTags(contentTags)))
+		return mr.mergeFile(ss, cb)
 	})
 }
 
-func applyFullDeletes(frs []*FileReader) []*FileReader {
-	var result []*FileReader
-	// Filter out the file readers with lower priority than the highest priority file reader that contains a full file deletion.
-	for _, fr := range frs {
-		result = append(result, fr)
+func (mr *MergeReader) mergeFile(ss []stream, cb func(*FileMergeReader) error, deleter ...func(string, ...string)) (retErr error) {
+	// Convert generic streams to file streams.
+	var fss []*fileStream
+	for _, s := range ss {
+		fss = append(fss, s.(*fileStream))
+	}
+	// Progress each stream and collect each file reader.
+	var i int
+	var frs []*FileReader
+	for ; i < len(fss); i++ {
+		fr, err := fss[i].r.Next()
+		if err != nil {
+			return err
+		}
+		frs = append(frs, fr)
+		// Filter out the file readers with lower priority than the highest priority file reader that contains a full file deletion.
 		if isFullDelete(fr.Index()) {
-			return result
+			break
+		}
+	}
+	// Progress each lower priority stream past the deleted file(s).
+	for i += 1; i < len(fss); i++ {
+		fr, err := fss[i].r.Next()
+		if err != nil {
+			return err
+		}
+		// If the full delete is for a directory, progress all lower priority streams past the directory.
+		if strings.HasSuffix(fr.Index().Path, "/") {
+			next := strings.TrimRight(fr.Index().Path, "/") + "0"
```

Comment:
This is a little sketchy. Unless we check all the paths to make sure that 0 is the first allowed character we should probably find a better way of doing this.  Maybe have a utility function that does this so the intent is more obvious.  A NULL character instead of 0 might make this slightly less sketchy.

Reply:
The immediate next character after "/" in a lexicograpical sort is "0", so the lowest next path (in terms of the sort) after "/dir/" would be "/dir0". The next in this case defines an exclusive upper bound for the iteration. I can put this in a utility function.